src/query/storage/m3/dynamic_cluster.go

```diff
@@ -88,10 +96,397 @@ func NewDynamicClusterOptions() DynamicClusterOptions {
+	clusterCfgs []DynamicClusterNamespaceConfiguration
+	logger      *zap.Logger
+	iOpts       instrument.Options
+
+	sync.RWMutex
+
+	namespaces              ClusterNamespaces
+	unaggregatedNamespace   ClusterNamespace
+	aggregatedNamespaces    map[RetentionResolution]ClusterNamespace
+	namespacesByEtcdCluster map[int]clusterNamespaceLookup
+
+	nsWatches []namespace.NamespaceWatch
+	closed    bool
+}
 
 // NewDynamicClusters creates an implementation of the Clusters interface
 // supports dynamic updating of cluster namespaces.
 func NewDynamicClusters(_ DynamicClusterOptions) (Clusters, error) {
 	return nil, errors.New("dynamic cluster configuration not yet supported")
 }
+
+// TODO(nate): Replace constructor above with this method
+// once namespace staging is complete.
+//
+// newDynamicClusters creates an implementation of the Clusters interface
+// supports dynamic updating of cluster namespaces.
+func newDynamicClusters(opts DynamicClusterOptions) (Clusters, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	cluster := dynamicCluster{
+		clusterCfgs:             opts.DynamicClusterNamespaceConfiguration(),
+		logger:                  opts.InstrumentOptions().Logger(),
+		iOpts:                   opts.InstrumentOptions(),
+		namespacesByEtcdCluster: make(map[int]clusterNamespaceLookup),
+	}
+
+	if err := cluster.init(); err != nil {
+		if cErr := cluster.Close(); cErr != nil {
+			return nil, cErr
+		}
+
+		return nil, err
+	}
+
+	return &cluster, nil
+}
+
+func (d *dynamicCluster) init() error {
+	d.logger.Info("creating namespaces watches", zap.Int("clusters", len(d.clusterCfgs)))
+
+	var (
+		wg       sync.WaitGroup
+		multiErr xerrors.MultiError
+		errLock  sync.Mutex
+	)
+	// Configure watch for each cluster provided
+	for i, cfg := range d.clusterCfgs {
+		i := i
+		cfg := cfg
+
+		wg.Add(1)
+		go func() {
+			if err := d.initNamespaceWatch(i, cfg); err != nil {
+				errLock.Lock()
+				multiErr = multiErr.Add(err)
+				errLock.Unlock()
+			}
+			wg.Done()
+		}()
+	}
+
+	wg.Wait()
+	if !multiErr.Empty() {
+		return multiErr.FinalError()
+	}
+
+	return nil
+}
+
+func (d *dynamicCluster) initNamespaceWatch(etcdClusterId int, cfg DynamicClusterNamespaceConfiguration) error {
+	registry, err := cfg.nsInitializer.Init()
+	if err != nil {
+		return err
+	}
+
+	// Get a namespace watch
+	watch, err := registry.Watch()
+	if err != nil {
+		return err
+	}
+
+	// Wait till first namespaces value is received and set the value.
+	d.logger.Info("resolving namespaces with namespace watch", zap.Int("cluster", etcdClusterId))
+	<-watch.C()
+
+	updater := func(namespaces namespace.Map) error {
+		d.updateNamespaces(etcdClusterId, cfg, namespaces)
+		return nil
+	}
+	nsWatch := namespace.NewNamespaceWatch(updater, watch, d.iOpts)
+	if err = nsWatch.Start(); err != nil {
+		return err
+	}
+
+	nsMap := watch.Get()
+	d.updateNamespaces(etcdClusterId, cfg, nsMap)
+
+	d.Lock()
+	d.nsWatches = append(d.nsWatches, nsWatch)
```

Comment:
Should there be a `Stop()`/`Close()` method that stops all of these watches exposed? Maybe not since there's not really a reason to deregister these right?